### ISSUE TITLE
HAR file can be read using HARLib but not parsed

### ISSUE THREAD

**Issue by neillfontes** [lightbody/browsermob-proxy#143]
I would like to capture all the requests and responses made by a Selenium test. I've started the proxy before the test and I have the HAR file ready in the end of the test. When I iterate over all the entries of the HAR file I need to parse the JSON into Java Objects using ObjectMapper or XStream. But when I run the parser over a given entry, the JSON format for that entry is invalid.

The JSON entry can be validated using: http://jsonlint.com/ 

This one is a sample: 

"request": { "method": "GET", "url": "http://ads.yahoo.com/cms/v1?esig=1%xxxxxxxxxxxx&nwid=xxxxxxxx&sigv=1", "httpVersion": "HTTP",   "cookies": [{ "name": "B", "value": "xxxxxxxx&b=3&s=9h", "path": "null", "domain": "null", "expires": "null", "httpOnly": null, "secure": null, "comment": "", }
]
,   "headers": [{ "name": "Host", "value": "ads.yahoo.com", "comment": "null",  }, { "name": "Accept", "value": "image/webp,*/*;q=0.8", "comment": "null",  }, { "name": "User-Agent", "value": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36", "comment": "null",  }, { "name": "Accept-Encoding", "value": "gzip, deflate, sdch", "comment": "null",  }, { "name": "Accept-Language", "value": "en-US,en;q=0.8", "comment": "null",  }, { "name": "Cookie", "value": "B=xxxxxxxxx&b=3&s=9h", "comment": "null",  }]
,   "queryString": [{ "name": "esig", "value": "1~xxxxxxxxxxxxxxxxx", "comment": "null",  }, { "name": "nwid", "value": "xxxxxxxx", "comment": "null",  }, { "name": "sigv", "value": "1", "comment": "null",  }]
, null,  "headersSize": 396,  "bodySize": 0,  "comment": "", }

JSON can be beautified with: http://jsonviewer.stack.hu/

Issues

1- The very last comma of each inner/root nodes are causing a validation failure
2- The "null," entry is causing a validation failure (could it be ignored?)

The HARLib in question is this one: https://sites.google.com/site/frogthinkerorg/projects/harlib, and I used XStream for parsing, without luck (unless doing some manual interventions on the JSON entry).

**Comment by jekh** [comment]
What version of BMP are you running? I haven't seen this happen with the latest beta 10 snapshot.

If you still see this with the latest code, I have an open branch to replace our very old version of Jackson with Jackson 2.4. I'm just waiting on some of the other PRs to be merged before I submit it.

**Comment by neillfontes** [comment]
Hello jekh,

I got browsermob-proxy-2.0-beta-9.jar. I will update the jar, re-run the test and post the output here.

**Comment by neillfontes** [comment]
Hello jekh,

I made a test this morning with beta 10 and selenium 2.44 (upgrade was seamless except for a few lib updates), but the issue persists. I'm still can't get the JSON automatically parsed without manual intervention. This is before and after the formatting:

Original request
"request": { "method": "GET", "url": "http://ads.yahoo.com/cms/v1?esig=1%7E7315a025058f3128185459bfaf16e164414683fc&nwid=10000545908&sigv=1", "httpVersion": "HTTP/1.1",   "cookies": [{ "name": "B", "value": "3j7fnthaava32&b=3&s=33", "path": "null", "domain": "null", "expires": "null", "httpOnly": null, "secure": null, "comment": "", }
]
,   "headers": [{ "name": "Host", "value": "ads.yahoo.com", "comment": "null",  }, { "name": "Accept", "value": "image/webp,*/*;q=0.8", "comment": "null",  }, { "name": "User-Agent", "value": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36", "comment": "null",  }, { "name": "Accept-Encoding", "value": "gzip, deflate, sdch", "comment": "null",  }, { "name": "Accept-Language", "value": "en-US,en;q=0.8", "comment": "null",  }, { "name": "Cookie", "value": "B=3j7fnthaava32&b=3&s=33", "comment": "null",  }]
,   "queryString": [{ "name": "esig", "value": "1~7315a025058f3128185459bfaf16e164414683fc", "comment": "null",  }, { "name": "nwid", "value": "10000545908", "comment": "null",  }, { "name": "sigv", "value": "1", "comment": "null",  }]
, null,  "headersSize": 392,  "bodySize": 0,  "comment": "", }

After fixing
{ "method": "GET", "url": "http://ads.yahoo.com/cms/v1?esig=1%7E7315a025058f3128185459bfaf16e164414683fc&nwid=10000545908&sigv=1", "httpVersion": "HTTP/1.1",   "cookies": [{ "name": "B", "value": "3j7fnthaava32&b=3&s=33", "path": "null", "domain": "null", "expires": "null", "httpOnly": "null", "secure": "null", "comment": ""}
]
,   "headers": [{ "name": "Host", "value": "ads.yahoo.com", "comment": "null"}, { "name": "Accept", "value": "image/webp,*/*;q=0.8", "comment": "null"}, { "name": "User-Agent", "value": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36", "comment": "null"}, { "name": "Accept-Encoding", "value": "gzip, deflate, sdch", "comment": "null"}, { "name": "Accept-Language", "value": "en-US,en;q=0.8", "comment": "null"}, { "name": "Cookie", "value": "B=3j7fnthaava32&b=3&s=33", "comment": "null"}]
,   "queryString": [{ "name": "esig", "value": "1~7315a025058f3128185459bfaf16e164414683fc", "comment": "null"}, { "name": "nwid", "value": "10000545908", "comment": "null"}, { "name": "sigv", "value": "1", "comment": "null"}]
,   "headersSize": 392,  "bodySize": 0,  "comment": ""}

**Comment by jekh** [comment]
I just tried hitting that URL with the latest beta 10 snapshot and did not run into this issue when retrieving the HAR from the proxy using the proxy's REST API. Since you're iterating over the HAR objects directly, I suspect the issue is the JSON library you're using. For one thing, it looks like you are using both jackson 1.9 and jackson 2, which is almost certainly not going to give you what you want. Additionally, BMP uses jackson 1.7 (~4 years old), which may be further confusing things.

The easiest thing to do would be to make your client code use jackson 1.x exclusively and retest.

@lightbody  -- my jackson 2 upgrade branch depends on PR #142. Could you review that and merge it if it looks good? (It's mostly adding tests, with a few changes to the Har* classes to make them threadsafe.) Then I'll squash my jackson upgrade commits and submit another PR to get us onto a current and supported version of jackson.

@neillfontes -- here's what I see using the proxy's REST API (after beautification):
`$ curl -X GET http://localhost:8080/proxy/8081/har`

```json
{
    "log": {
        "browser": {
            "name": "Firefox",
            "comment": "",
            "version": "34.0"
        },
        "pages": [{
            "pageTimings": {
                "comment": ""
            },
            "startedDateTime": "2015-01-09T18:06:54.717-08:00",
            "title": "",
            "id": "Page 1",
            "comment": ""
        }],
        "creator": {
            "name": "BrowserMob Proxy",
            "comment": "",
            "version": "2.0"
        },
        "entries": [{
            "pageref": "Page 1",
            "serverIPAddress": "216.39.55.13",
            "startedDateTime": "2015-01-09T18:08:07.348-08:00",
            "timings": {
                "blocked": 26,
                "dns": 79,
                "connect": 1,
                "ssl": 0,
                "send": 3,
                "wait": 82,
                "receive": 20,
                "comment": ""
            },
            "time": 185,
            "cache": {},
            "comment": "",
            "request": {
                "httpVersion": "HTTP/1.1",
                "headersSize": 401,
                "bodySize": 0,
                "queryString": [{
                    "name": "esig",
                    "value": "1~7315a025058f3128185459bfaf16e164414683fc"
                }, {
                    "name": "nwid",
                    "value": "10000545908"
                }, {
                    "name": "sigv",
                    "value": "1"
                }],
                "method": "GET",
                "comment": "",
                "headers": [],
                "url": "http://ads.yahoo.com/cms/v1?esig=1%7E7315a025058f3128185459bfaf16e164414683fc&nwid=10000545908&sigv=1",
                "cookies": [{
                    "expires": null,
                    "name": "B",
                    "value": "7cjtu35a2jaj0&b=3&s=bu",
                    "comment": ""
                }]
            },
            "response": {
                "httpVersion": "HTTP/1.1",
                "headersSize": 510,
                "statusText": "Found",
                "redirectURL": "",
                "bodySize": 0,
                "content": {
                    "size": 0,
                    "comment": "",
                    "mimeType": "text/plain; charset=utf-8"
                },
                "comment": "",
                "headers": [],
                "status": 302,
                "cookies": []
            }
        }, {
            "pageref": "Page 1",
            "serverIPAddress": "74.119.117.94",
            "startedDateTime": "2015-01-09T18:08:07.539-08:00",
            "timings": {
                "blocked": 0,
                "dns": 3,
                "connect": 1,
                "ssl": 0,
                "send": 0,
                "wait": 96,
                "receive": 3,
                "comment": ""
            },
            "time": 103,
            "cache": {},
            "comment": "",
            "request": {
                "httpVersion": "HTTP/1.1",
                "headersSize": 353,
                "bodySize": 0,
                "queryString": [{
                    "name": "xid",
                    "value": "uvVHgKRuVaCZa3ffr8k3bdHL"
                }],
                "method": "GET",
                "comment": "",
                "headers": [],
                "url": "http://dis.criteo.com/dis/rtb/rightmedia/cookiematch.aspx?xid=uvVHgKRuVaCZa3ffr8k3bdHL",
                "cookies": []
            },
            "response": {
                "httpVersion": "HTTP/1.1",
                "headersSize": 303,
                "statusText": "OK",
                "redirectURL": "",
                "bodySize": 43,
                "content": {
                    "size": 43,
                    "comment": "",
                    "mimeType": "image/gif"
                },
                "comment": "",
                "headers": [],
                "status": 200,
                "cookies": []
            }
        }],
        "comment": "",
        "version": "1.2"
    }
}
```

**Comment by jekh** [comment]
As of now, master is now on Jackson 2. You pull the latest code and give this another try. Just make sure you're not trying to use Jackson 1 to serialize the HAR objects anymore.

**Comment by neillfontes** [comment]
Hello Jekh,

I will make another test and I will post the results here.

Thanks,

Neill

**Comment by claytonneal** [comment]
Hello.
Im using HARLib. I installed the jar without the Jackson parser, but put the latest Jackson reference in my pom. The error i get is:
Missing text field in postData element
--> Did you solve this ?

**Comment by claytonneal** [comment]
I switched from HARLib to this project, which is using updated Jackson - and for me its working
https://github.com/sdstoehr/har-reader

**Comment by jekh** [comment]
I'm going to close this as resolved by the Jackson 2 update.
